### PR TITLE
[SM-429] Add permission checks to access policy endpoints

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to add permission checks to access policy endpoints and add a new endpoint to get potential grantees.

## Code changes

- bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/CreateAccessPoliciesCommand.cs
src/Core/SecretsManager/Commands/AccessPolicies/Interfaces/ICreateAccessPoliciesCommand.cs

Add in permission check and return the project's entire access policy view instead of just the newly created policies.

- bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs
src/Core/SecretsManager/Commands/AccessPolicies/Interfaces/IDeleteAccessPolicyCommand.cs

Add in permission check

- bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/UpdateAccessPolicyCommand.cs
src/Core/SecretsManager/Commands/AccessPolicies/Interfaces/IUpdateAccessPolicyCommand.cs

Add in permission check

- bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ServiceAccountRepository.cs
src/Core/SecretsManager/Repositories/IServiceAccountRepository.cs

Make a new method for getting service accounts that could potentially be applied to project access policiy grants.

- bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/CreateAccessPoliciesCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/DeleteAccessPolicyCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/UpdateAccessPolicyCommandTests.cs

Test permissions in unit tests.

- bitwarden_license/test/Commercial.Core.Test/SecretsManager/Enums/AccessPolicyType.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/Enums/PermissionType.cs
test/Api.IntegrationTest/SecretsManager/Enums/PermissionType.cs

Enums to simplify test cases

- src/Api/SecretsManager/Controllers/AccessPoliciesController.cs

Add in permission checks and two new endpoints to get a project's potential grantees.

- src/Api/SecretsManager/Models/Response/PotentialGranteeResponseModel.cs

New response model for pontential grantees.

- test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs

Integration tests for permission changes.

- test/Api.Test/SecretsManager/Controllers/AccessPoliciesControllerTests.cs

Test permissions in unit tests.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/CreateAccessPoliciesCommand.cs [hunk 1]
```diff
@@ -39,7 +71,7 @@ public async Task<List<BaseAccessPolicy>> CreateAsync(List<BaseAccessPolicy> acc
                 throw new BadRequestException("Resource already exists");
             }
         }
-
-        return await _accessPolicyRepository.CreateManyAsync(accessPolicies);
+        await _accessPolicyRepository.CreateManyAsync(accessPolicies);
+        return await _accessPolicyRepository.GetManyByProjectId(projectId);
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/CreateAccessPoliciesCommand.cs:75]: Any reason we need to re-fetch them?

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs [hunk 2]
```diff
@@ -1,27 +1,135 @@
...
             throw new NotFoundException();
         }
 
+        switch (accessPolicy)
+        {
+            case UserProjectAccessPolicy ap:
+                if (ap.GrantedProjectId == null)
+                {
+                    throw new BadRequestException();
+                }
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs:44]: Can this even happen?

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs [hunk 3]
```diff
@@ -1,27 +1,135 @@
...
+
+                var project = await _projectRepository.GetByIdAsync(ap.GrantedProjectId.Value);
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs:46]: Any reason we don't include the project in the access policy query?

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs [hunk 4]
```diff
@@ -1,27 +1,135 @@
...
+    }
+
+    private async Task DeleteServiceAccountGrantAsync(Guid id, Guid userId, Guid? serviceAccountId)
+    {
+        if (serviceAccountId == null)
+        {
+            throw new BadRequestException();
+        }
+
+        var serviceAccount = await _serviceAccountRepository.GetByIdAsync(serviceAccountId.Value);
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs:84]: Could we just include the service account in the access policy directly? Should let us simplify this as well. We could move the actual access  check into the CheckPermissions, and move out the delete async into the regular method.

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/UpdateAccessPolicyCommand.cs [hunk 5]
```diff
@@ -8,24 +10,135 @@ namespace Bit.Commercial.Core.SecretsManager.Commands.AccessPolicies;
...
 
+        switch (accessPolicy)
+        {
+            case UserProjectAccessPolicy ap:
+                if (ap.GrantedProjectId == null)
+                {
+                    throw new NotFoundException();
+                }
+
+                var project = await _projectRepository.GetByIdAsync(ap.GrantedProjectId.Value);
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/UpdateAccessPolicyCommand.cs:45]: Same comment as Delete, if we include the related objects directly we can simplify this quite a bit.

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ServiceAccountRepository.cs [hunk 6]
```diff
@@ -54,6 +54,23 @@ public async Task<bool> UserHasWriteAccessToServiceAccount(Guid id, Guid userId)
         return await query.AnyAsync();
     }
 
+    public async Task<IEnumerable<Core.SecretsManager.Entities.ServiceAccount>> GetPotentialGranteesAsync(Guid organizationId, Guid userId, AccessClientType accessType)
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ServiceAccountRepository.cs:57]: The method name is fairly non descriptive. I believe this method is identical to `GetManyByOrganizationIdAsync` except we check write access instead?

`GetManyByOrganizationIdWriteAccessAsync` maybe.

File: src/Api/SecretsManager/Controllers/AccessPoliciesController.cs [hunk 7]
```diff
@@ -12,57 +18,142 @@ namespace Bit.Api.SecretsManager.Controllers;
...
+        _groupRepository = groupRepository;
+        _organizationUserRepository = organizationUserRepository;
         _accessPolicyRepository = accessPolicyRepository;
         _createAccessPoliciesCommand = createAccessPoliciesCommand;
         _deleteAccessPolicyCommand = deleteAccessPolicyCommand;
         _updateAccessPolicyCommand = updateAccessPolicyCommand;
     }
 
-    [HttpPost("/projects/{id}/access-policies")]
+    [HttpPost("/projects/{id:guid}/access-policies")]
```
Comment (Hinton) [inline src/Api/SecretsManager/Controllers/AccessPoliciesController.cs:55]: There's a discussion on slack regarding specifying `guid` in the urls. We decided against doing so.

**Review by Hinton — COMMENTED** [review]
Initial review, I haven't gone through the tests in detail yet.

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/CreateAccessPoliciesCommand.cs (hunk 1 above)
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/CreateAccessPoliciesCommand.cs:75]: `await _accessPolicyRepository.CreateManyAsync(accessPolicies);` just returns what was created. 

Since this endpoint is creating for projects.  `[HttpPost("/projects/{id}/access-policies")]` 
It made more sense from an API perspective and consuming it from the web project that the whole project access policies view was returned.

I'm open to changing it if required.

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs (hunk 3 above)
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs]: Great suggestion, really like it. 

Resolved in https://github.com/bitwarden/server/pull/2628/commits/cd92ea867e0c629cdebeccbb8a3d9499053cef04

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs (hunk 4 above)
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/cd92ea867e0c629cdebeccbb8a3d9499053cef04

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/UpdateAccessPolicyCommand.cs (hunk 5 above)
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/UpdateAccessPolicyCommand.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/cd92ea867e0c629cdebeccbb8a3d9499053cef04

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ServiceAccountRepository.cs (hunk 6 above)
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ServiceAccountRepository.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/cd92ea867e0c629cdebeccbb8a3d9499053cef04

File: src/Api/SecretsManager/Controllers/AccessPoliciesController.cs [hunk 8]
```diff
@@ -12,16 +18,34 @@ namespace Bit.Api.SecretsManager.Controllers;
...
         ICreateAccessPoliciesCommand createAccessPoliciesCommand,
         IDeleteAccessPolicyCommand deleteAccessPolicyCommand,
         IUpdateAccessPolicyCommand updateAccessPolicyCommand)
     {
+        _userService = userService;
+        _serviceAccountRepository = serviceAccountRepository;
+        _projectRepository = projectRepository;
+        _currentContext = currentContext;
+        _groupRepository = groupRepository;
+        _organizationUserRepository = organizationUserRepository;
```
Comment (Hinton) [inline src/Api/SecretsManager/Controllers/AccessPoliciesController.cs:48]: Let's group repositories and services.

```suggestion
        _userService = userService;
        _currentContext = currentContext;
        _serviceAccountRepository = serviceAccountRepository;
        _projectRepository = projectRepository;
        _groupRepository = groupRepository;
        _organizationUserRepository = organizationUserRepository;
```

File: src/Api/SecretsManager/Controllers/AccessPoliciesController.cs [hunk 9]
```diff
@@ -48,21 +76,84 @@ public async Task<ProjectAccessPoliciesResponseModel> GetProjectAccessPoliciesAs
...
 
     [HttpDelete("{id}")]
     public async Task DeleteAccessPolicyAsync([FromRoute] Guid id)
     {
-        await _deleteAccessPolicyCommand.DeleteAsync(id);
+        var userId = _userService.GetProperUserId(User).Value;
+        await _deleteAccessPolicyCommand.DeleteAsync(id, userId);
+    }
+
+    [HttpGet("/projects/{id}/access-policies/people/potential-grantees")]
```
Comment (Hinton) [inline src/Api/SecretsManager/Controllers/AccessPoliciesController.cs:99]: Are these related to projects somehow? Wondering if we should make two generic endpoints that we re-use for all entities?

File: bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/DeleteAccessPolicyCommandTests.cs [hunk 10]
```diff
@@ -11,28 +15,241 @@
...
+            case AccessPolicyType.ServiceAccountProjectAccessPolicy:
+                policyToReturn = new ServiceAccountProjectAccessPolicy
+                {
+                    Id = data,
+                    GrantedProjectId = grantedProject.Id,
+                    ServiceAccount = mockServiceAccount,
+                    GrantedProject = grantedProject,
+                };
+                break;
+        }
```
Comment (Hinton) [inline bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/DeleteAccessPolicyCommandTests.cs:138]: I think we can clean up the tests a bit by extracting this into a function? It seems to be mostly reused in `DeleteAccessPolicy_ProjectGrants_PermissionsCheck_Success`.

File: bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/UpdateAccessPolicyCommandTests.cs [hunk 11]
```diff
@@ -11,30 +15,290 @@
...
+        {
+            case PermissionType.RunAsAdmin:
+                sutProvider.GetDependency<ICurrentContext>().OrganizationAdmin(grantedProject.OrganizationId)
+                    .Returns(true);
+                break;
+            case PermissionType.RunAsUserWithPermission:
+                sutProvider.GetDependency<IProjectRepository>().UserHasWriteAccessToProject(grantedProject.Id, userId)
+                    .Returns(true);
+                break;
+        }
```
Comment (Hinton) [inline bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/UpdateAccessPolicyCommandTests.cs:112]: Extract to a function? Re-used in `UpdateAsync_ServiceAccountGrants_PermissionsCheck_Success`. Would also make the test more easily digested.

File: bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/UpdateAccessPolicyCommandTests.cs [hunk 12]
```diff
@@ -11,30 +15,290 @@
...
+                    {
+                        Id = data,
+                        GrantedServiceAccountId = grantedServiceAccount.Id,
+                        Read = true,
+                        Write = true,
+                        Group = mockGroup,
+                        GrantedServiceAccount = grantedServiceAccount,
+                    };
+                break;
+        }
```
Comment (Hinton) [inline bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/UpdateAccessPolicyCommandTests.cs:231]: Extract to function?

File: bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/DeleteAccessPolicyCommandTests.cs [hunk 13]
```diff
@@ -11,28 +15,241 @@
...
+                policyToReturn =
+                    new GroupServiceAccountAccessPolicy
+                    {
+                        Id = data,
+                        GrantedServiceAccountId = grantedServiceAccount.Id,
+                        Group = mockGroup,
+                        GrantedServiceAccount = grantedServiceAccount,
+                    };
+                break;
+        }
```
Comment (Hinton) [inline bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/DeleteAccessPolicyCommandTests.cs:187]: Extract?

File: bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/DeleteAccessPolicyCommandTests.cs [hunk 14]
```diff
@@ -11,28 +15,241 @@
...
+        {
+            case PermissionType.RunAsAdmin:
+                sutProvider.GetDependency<ICurrentContext>().OrganizationAdmin(grantedProject.OrganizationId)
+                    .Returns(true);
+                break;
+            case PermissionType.RunAsUserWithPermission:
+                sutProvider.GetDependency<IProjectRepository>().UserHasWriteAccessToProject(grantedProject.Id, userId)
+                    .Returns(true);
+                break;
+        }
```
Comment (Hinton) [inline bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/DeleteAccessPolicyCommandTests.cs:94]: Extract?

File: test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs [hunk 15]
```diff
@@ -35,37 +38,96 @@ public AccessPoliciesControllerTest(ApiApplicationFactory factory)
...
+    [InlineData(true, false)]
+    [InlineData(false, true)]
+    public async Task CreateProjectAccessPolicies_SmNotEnabled_NotFound(bool useSecrets, bool accessSecrets)
+    {
+        var (org, _) = await _organizationHelper.Initialize(useSecrets, accessSecrets);
+        await LoginAsync(_email);
 
-    [Fact]
-    public async Task CreateProjectAccessPolicies()
+        var initialProject = await _projectRepository.CreateAsync(new Project
```
Comment (Hinton) [inline test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs:67]: Personal opinion, but unless we update the project we should name it `project` instead of `initialProject`.

```suggestion
        var project = await _projectRepository.CreateAsync(new Project
```

File: test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs [hunk 16]
```diff
@@ -91,9 +153,100 @@ public async Task CreateProjectAccessPolicies()
...
-    public async Task UpdateAccessPolicy()
+    public async Task CreateProjectAccessPolicies_NoPermission()
     {
-        var initData = await SetupAccessPolicyRequest();
+        // Create a new account as a user
+        var (org, _) = await _organizationHelper.Initialize(true, true);
+        var (email, orgUser) = await _organizationHelper.CreateNewUser(OrganizationUserType.User, true);
+        await LoginAsync(email);
+
+        var initialProject = await _projectRepository.CreateAsync(new Project
```
Comment (Hinton) [inline test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs:163]:
```suggestion
        var project = await _projectRepository.CreateAsync(new Project
```

File: test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs [hunk 17]
```diff
@@ -91,9 +153,100 @@ public async Task CreateProjectAccessPolicies()
...
+        {
+            OrganizationId = org.Id,
+            Name = _mockEncryptedString,
+        });
+
+        var initialServiceAccount = await _serviceAccountRepository.CreateAsync(new ServiceAccount
```
Comment (Hinton) [inline test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs:169]:
```suggestion
        var serviceAccount = await _serviceAccountRepository.CreateAsync(new ServiceAccount
```

**Review by Hinton — COMMENTED** [review]
Went through everything this time. But mostly skimmed the tests. Looks good, some feedback and then we should be good.

File: src/Api/SecretsManager/Controllers/AccessPoliciesController.cs (hunk 9 above)
Comment (Thomas-Avery) [reply src/Api/SecretsManager/Controllers/AccessPoliciesController.cs:99]: They're related to projects in the sense that we check if the user has write permissions to the project to fetch the potential grantees. 

Another option would be 

`[HttpGet("/organizations/{id}/access-policies/people/potential-grantees")]`
`[HttpGet("/organizations/{id}/access-policies/service-accounts/potential-grantees")]`

and just check that the user has secrets manager access. 

The reason I went for this approach was because the user endpoint we were using previously was more restrictive. 

Thoughts on allowing any secret's manager user in the org to hit the endpoint? 

To note the service account potential grantees, fetch checks that the user has write permission to each service account returned.

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs [hunk 18]
```diff
@@ -1,27 +1,139 @@
...
         }
 
-        await _accessPolicyRepository.DeleteAsync(id);
+        switch (accessPolicy)
+        {
+            case UserProjectAccessPolicy ap:
+                if (ap.GrantedProject == null)
+                {
+                    throw new NotFoundException();
+                }
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs:44]: The check is there because it's a nullable field, and we get warnings if we don't check for null.

It's nullable because when we are mapping requests onto the model we won't populate the field. 

When mapping database queries, the field should never be null.

Not sure what you mean by a validate method.

File: src/Api/SecretsManager/Controllers/AccessPoliciesController.cs (hunk 9 above)
Comment (Hinton) [reply src/Api/SecretsManager/Controllers/AccessPoliciesController.cs:99]: We could check if the user has write access to any project or service account. But not sure if it's worth it.

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (Hinton) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs:44]: Oh right. You can ignore the warnings using `!`, or null coalece with `?`.

File: test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs (hunk 17 above)
Comment (Thomas-Avery) [reply test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/79bce819b50d75b5ce021b4b4dfad21dc9c97cd8

File: test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs (hunk 16 above)
Comment (Thomas-Avery) [reply test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/79bce819b50d75b5ce021b4b4dfad21dc9c97cd8

File: test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs (hunk 15 above)
Comment (Thomas-Avery) [reply test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/79bce819b50d75b5ce021b4b4dfad21dc9c97cd8

File: bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/DeleteAccessPolicyCommandTests.cs (hunk 14 above)
Comment (Thomas-Avery) [reply bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/DeleteAccessPolicyCommandTests.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/79bce819b50d75b5ce021b4b4dfad21dc9c97cd8

File: bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/DeleteAccessPolicyCommandTests.cs (hunk 13 above)
Comment (Thomas-Avery) [reply bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/DeleteAccessPolicyCommandTests.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/79bce819b50d75b5ce021b4b4dfad21dc9c97cd8

File: bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/UpdateAccessPolicyCommandTests.cs (hunk 12 above)
Comment (Thomas-Avery) [reply bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/UpdateAccessPolicyCommandTests.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/79bce819b50d75b5ce021b4b4dfad21dc9c97cd8

File: bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/UpdateAccessPolicyCommandTests.cs (hunk 11 above)
Comment (Thomas-Avery) [reply bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/UpdateAccessPolicyCommandTests.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/79bce819b50d75b5ce021b4b4dfad21dc9c97cd8

File: bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/DeleteAccessPolicyCommandTests.cs (hunk 10 above)
Comment (Thomas-Avery) [reply bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/DeleteAccessPolicyCommandTests.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/79bce819b50d75b5ce021b4b4dfad21dc9c97cd8

File: src/Api/SecretsManager/Controllers/AccessPoliciesController.cs (hunk 8 above)
Comment (Thomas-Avery) [reply src/Api/SecretsManager/Controllers/AccessPoliciesController.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/79bce819b50d75b5ce021b4b4dfad21dc9c97cd8

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/UpdateAccessPolicyCommand.cs [hunk 19]
```diff
@@ -8,24 +10,132 @@ namespace Bit.Commercial.Core.SecretsManager.Commands.AccessPolicies;
...
-    public async Task<BaseAccessPolicy> UpdateAsync(Guid id, bool read, bool write)
+    public async Task<BaseAccessPolicy> UpdateAsync(Guid id, bool read, bool write, Guid userId)
     {
         var accessPolicy = await _accessPolicyRepository.GetByIdAsync(id);
         if (accessPolicy == null)
         {
             throw new NotFoundException();
         }
 
+        switch (accessPolicy)
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/UpdateAccessPolicyCommand.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/79bce819b50d75b5ce021b4b4dfad21dc9c97cd8

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs [hunk 20]
```diff
@@ -1,27 +1,139 @@
...
+    public async Task DeleteAsync(Guid id, Guid userId)
     {
         var accessPolicy = await _accessPolicyRepository.GetByIdAsync(id);
         if (accessPolicy == null)
         {
             throw new NotFoundException();
         }
 
-        await _accessPolicyRepository.DeleteAsync(id);
+        switch (accessPolicy)
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs]: I've got it cleaned up in https://github.com/bitwarden/server/pull/2628/commits/79bce819b50d75b5ce021b4b4dfad21dc9c97cd8


How things are currently setup, we wouldn't be able to pass in accessClient. 

Creating the access client requires knowledge of what organization we are targeting. 

It isn't until we preform the pattern matching on the access policy that we have access to the target organization.

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs [hunk 21]
```diff
@@ -1,27 +1,139 @@
...
+                break;
+            case GroupServiceAccountAccessPolicy ap:
+                if (ap.GrantedServiceAccount == null)
+                {
+                    throw new NotFoundException();
+                }
+
+                await CheckPermissionsAsync(ap.GrantedServiceAccount.OrganizationId, userId,
+                    serviceAccountIdToCheck: ap.GrantedServiceAccountId);
+                await _accessPolicyRepository.DeleteAsync(id);
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/79bce819b50d75b5ce021b4b4dfad21dc9c97cd8

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/79bce819b50d75b5ce021b4b4dfad21dc9c97cd8

File: src/Api/SecretsManager/Controllers/AccessPoliciesController.cs (hunk 9 above)
Comment (Thomas-Avery) [reply src/Api/SecretsManager/Controllers/AccessPoliciesController.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/cc6f876b03a420ebe1f0f44d135825de6f900dab

**Review by Hinton — DISMISSED** [review]
Minor nitpick.

File: src/Api/SecretsManager/Controllers/AccessPoliciesController.cs [hunk 22]
```diff
@@ -48,21 +76,83 @@ public async Task<ProjectAccessPoliciesResponseModel> GetProjectAccessPoliciesAs
...
     [HttpDelete("{id}")]
     public async Task DeleteAccessPolicyAsync([FromRoute] Guid id)
     {
-        await _deleteAccessPolicyCommand.DeleteAsync(id);
+        var userId = _userService.GetProperUserId(User).Value;
+        await _deleteAccessPolicyCommand.DeleteAsync(id, userId);
+    }
+
+    [HttpGet("/organizations/{id}/access-policies/potential-grantees")]
+    public async Task<ListResponseModel<PotentialGranteeResponseModel>> GetPotentialGranteesAsync([FromRoute] Guid id, [FromQuery] bool includeServiceAccounts = false)
```
Comment (Hinton) [inline src/Api/SecretsManager/Controllers/AccessPoliciesController.cs:100]: This seems a bit odd, do we want two different endpoints for this? Or is there a scenario where you want both user/groups and service accounts?

Comment (Thomas-Avery) [reply src/Api/SecretsManager/Controllers/AccessPoliciesController.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/c976495d958eef84b4cd77ce057ba69e203760f6

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/UpdateAccessPolicyCommand.cs [hunk 23]
```diff
@@ -8,25 +10,102 @@ namespace Bit.Commercial.Core.SecretsManager.Commands.AccessPolicies;
...
+        {
+            case AccessClientType.NoAccessCheck:
+                hasAccess = true;
+                break;
+            case AccessClientType.User:
+                if (projectIdToCheck != null)
+                {
+                    hasAccess = await _projectRepository.UserHasWriteAccessToProject(projectIdToCheck.Value, userId);
+                }
+                else if (serviceAccountIdToCheck != null)
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/UpdateAccessPolicyCommand.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/c976495d958eef84b4cd77ce057ba69e203760f6

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs [hunk 24]
```diff
@@ -1,27 +1,107 @@
...
+        {
+            case AccessClientType.NoAccessCheck:
+                hasAccess = true;
+                break;
+            case AccessClientType.User:
+                if (projectIdToCheck != null)
+                {
+                    hasAccess = await _projectRepository.UserHasWriteAccessToProject(projectIdToCheck.Value, userId);
+                }
+                else if (serviceAccountIdToCheck != null)
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/c976495d958eef84b4cd77ce057ba69e203760f6

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs [hunk 25]
```diff
@@ -1,27 +1,107 @@
...
+        var accessClient = AccessClientHelper.ToAccessClient(_currentContext.ClientType, orgAdmin);
+
+        bool hasAccess;
+        switch (accessClient)
+        {
+            case AccessClientType.NoAccessCheck:
+                hasAccess = true;
+                break;
+            case AccessClientType.User:
+                if (projectIdToCheck != null)
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/DeleteAccessPolicyCommand.cs]: Resolved in https://github.com/bitwarden/server/pull/2628/commits/c976495d958eef84b4cd77ce057ba69e203760f6

**Review by Hinton — APPROVED** [review]
